File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import (
	"errors"
)

func CollatzConjecture(N int) (int, error) {
	if N <= 0 {
		err := errors.New("input N should be > 0.")
		return 0, err
	}
	steps := 0
	for {
		if steps > 1000 {
			break
		}
		if N <= 1 {
			break
		}
		steps++
		if (N % 2) == 0 {
			N = N / 2
			continue
		} else {
			N = (N * 3) + 1
			continue
		}
	}
	return steps, nil
}

File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  require Integer

  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()

  def calc(input) when input > 0 and is_integer(input) do
    calc(input, 0)
  end

  defp calc(1, steps) do
    steps
  end

  defp calc(input, steps) when input > 0 do
    if Integer.is_even(input) do
      calc(Kernel.div(input, 2), steps + 1)
    else
      calc(input * 3 + 1, steps + 1)
    end
  end
end
